base-containers/base/bin/_run_student_intern

```diff
@@ -36,96 +77,171 @@ def recv_fds(sock, msglen, maxfds):
+        message_size = struct.pack('!I', len(message))
+        client.send(message_size)
+        client.send(message)
+    else:
+        msg = {"type": "ssh_student", "ssh_user": ssh_user, "ssh_key": password,
+               "container_id": container_id}
+        loop = zmq.asyncio.ZMQEventLoop()
+        asyncio.set_event_loop(loop)
+        loop.run_until_complete(send_ssh_helper(msg))
+    # Wait for user to connect and leave
+    ssh_retval = ssh_wait(ssh_user)
+    return ssh_retval
+
+
+
+
+######################################################################
+######################################################################
+########### BODY #####################################################
+######################################################################
+######################################################################
+
+# Check the runtimes ######################################
 runtime = sys.argv[1]
-if runtime == "kata-runtime":
-    shared_kernel = False
-else:
-    shared_kernel = True
-    os.mkdir("/.__input")
-    shared_kernel_file = open("/.__input/__shared_kernel", "w")
-    shared_kernel_file.close()
-
-
-# Connect to the socket
-client = socket.socket(socket.AF_UNIX)  # , socket.SOCK_CLOEXEC) # for linux only
-client.connect("/__parent.sock")
-
-# Say hello
-print("Saying hello")
-client.send(b'H')
-print("Said hello")
-
-# Receive fds
-print("Receiving fds")
-msg, fds = recv_fds(client, 1, 3)
-assert msg == b'S'
-print("Received fds")
-
-# Unpack the start message # TODO: it's ugly
-print("Unpacking start cmd")
-unpacker = msgpack.Unpacker()
-start_cmd = None
-while start_cmd is None:
-    s = client.recv(1)
-    unpacker.feed(s)
-    for obj in unpacker:
-        start_cmd = obj
-print("Unpacked start cmd")
-user = start_cmd["user"]  # Know if you are worker (on docker) or root (on Kata)
+parent_runtime = sys.argv[2]
+shared_kernel, only_dockers = check_runtimes(runtime, parent_runtime)
+
+# Get the command to be run ##################################
+client, fds, start_cmd = receive_initial_command(only_dockers)
 
+# Check security
+student_container_id = start_cmd["student_container_id"]
+user = start_cmd["user"]  # Know if you are worker (on docker) or root (on Kata)
 if user == "root" and shared_kernel:
     print("You can not run as root on docker runtime")
     exit(251)
 
-
 # Add some elements to /etc/hosts and /etc/resolv.conf if needed
-system_files = {"hosts": ("/etc/hosts", True), "resolv.conf": ("/etc/resolv.conf", False)}
-for name, (spath, append) in system_files.items():
-    if os.path.exists(os.path.join('/task/systemfiles/', name)):
-        try:
-            open(spath, 'ab' if append else 'wb').write(b'\n' + open(os.path.join('/task/systemfiles/', name), 'rb').read())
-        except IOError:
-            exit(254)
+    system_files = {"hosts": ("/etc/hosts", True), "resolv.conf": ("/etc/resolv.conf", False)}
```

Comment:
These indents should probably not be there

Reply:
Indeed, my bad !